File: typescript/rna-transcription/rna-transcription.ts
class Transcriptor {

    static readonly NUCLEOTIDE_COMPLEMENTS = new Map<string, string>([
        ['G', 'C'],
        ['C', 'G'],
        ['T', 'A'],
        ['A', 'U']
    ]);

    public toRna(strand: string) {
        if (this.invalidInputData(strand)) {
            throw 'Invalid input DNA.'
        }
        return this.transcribeStrand(strand)
    }

    private transcribeStrand(strand: string) {
        return strand.split('').map(nucleotide => this.transcribeNucleotide(nucleotide)).join('')
    }

    private transcribeNucleotide(nucleotide: string) {
        return Transcriptor.NUCLEOTIDE_COMPLEMENTS.get(nucleotide)
    }

    private invalidInputData(strand: string): boolean {
        for (let nucleotide of strand) {
            if (!Transcriptor.NUCLEOTIDE_COMPLEMENTS.has(nucleotide)) {
                return true
            }
        }
        return false
    }
}

export default Transcriptor

File: go/rna-transcription/rna_transcription.go
package strand

var dnaToRna = map[rune]rune{
	'C': 'G',
	'G': 'C',
	'T': 'A',
	'A': 'U',
}

func ToRNA(dna string) (rna string) {
	for _, r := range dna {
		rna += string(dnaToRna[r])
	}
	return rna
}
